### ISSUE TITLE
How to use gt center or gt extreme and so on for testing?

### ISSUE THREAD

**Issue by moothes** [xingyizhou/ExtremeNet#11]
In the bottom of table 1, it shows the performance of testing with gt center and gt extreme and so on. How to implement that in your code?

**Comment by xingyizhou** [comment]
Hi,
  Thanks for your interests in our work. My implementation is a little bit hacky and I haven't included it in this repo. You will need to modify many files. Basically, you will need to pass flags in `test/coco_extreme.py` and `nnet/py_factory` and call a new forward pass in `models/py_utils/exkp.py` like the following (not cleaned-up version):
~~~
    def gt_validate(self, xs, targets, **kwargs):
        image = xs[0]
        gt_t_heat = targets[0]
        gt_l_heat = targets[1]
        gt_b_heat = targets[2]
        gt_r_heat = targets[3]
        gt_ct_heat = targets[4]
        gt_t_regr = targets[5]
        gt_l_regr = targets[6]
        gt_b_regr = targets[7]
        gt_r_regr = targets[8]

        inter = self.pre(image)
        outs  = []

        layers = zip(
            self.kps, self.cnvs,
            self.t_heats, self.l_heats, self.b_heats, self.r_heats,
            self.ct_heats,
            self.t_regrs, self.l_regrs, self.b_regrs, self.r_regrs,
            self.ct_regrs
        )
        for ind, layer in enumerate(layers):
            kp_, cnv_                          = layer[0:2]
            t_heat_, l_heat_, b_heat_, r_heat_ = layer[2:6]
            ct_heat_                           = layer[6]
            t_regr_, l_regr_, b_regr_, r_regr_ = layer[7:11]
            ct_regr_                           = layer[11]

            kp  = kp_(inter)
            cnv = cnv_(kp)

            if ind == self.nstack - 1:
                t_heat, l_heat = t_heat_(cnv), l_heat_(cnv)
                b_heat, r_heat = b_heat_(cnv), r_heat_(cnv)
                ct_heat        = ct_heat_(cnv)

                t_regr, l_regr = t_regr_(cnv), l_regr_(cnv)
                b_regr, r_regr = b_regr_(cnv), r_regr_(cnv)
                ct_regr        = ct_regr_(cnv)

                # outs += [t_heat, l_heat, b_heat, r_heat, ct_heat,
                #          t_regr, l_regr, b_regr, r_regr, ct_regr]
                # outs += [gt_t_heat, gt_l_heat, gt_b_heat, gt_r_heat, ct_heat,
                #          t_regr, l_regr, b_regr, r_regr, ct_regr]
                # outs += [t_heat, l_heat, b_heat, r_heat, gt_ct_heat,
                #          t_regr, l_regr, b_regr, r_regr, ct_regr]
                # outs += [gt_t_heat, gt_l_heat, gt_b_heat, gt_r_heat, gt_ct_heat,
                #          t_regr, l_regr, b_regr, r_regr, ct_regr]
                outs += [gt_t_heat, gt_l_heat, gt_b_heat, gt_r_heat, gt_ct_heat,
                         gt_t_regr, gt_l_regr, gt_b_regr, gt_r_regr, None]
                if DEBUG:
                    _debug(image, outs[0], outs[1], outs[2], outs[3], outs[4])
                    #  _debug_regr(image, t_regr, l_regr, b_regr, r_regr, ct_regr)

            if ind < self.nstack - 1:
                inter = self.inters_[ind](inter) + self.cnvs_[ind](cnv)
                inter = self.relu(inter)
                inter = self.inters[ind](inter)

        return self._decode(*outs[-10:], **kwargs), None

~~~

**Comment by moothes** [comment]
Thanks for your reply. 
By the way, I think my neg_loss implement is faster than yours:

    def _neg_loss(preds, gt):

        pos_inds = gt.eq(1).float()
        neg_inds = gt.lt(1).float()    
        num_pos  = pos_inds.float().sum()
        neg_weights = torch.pow(1 - gt, 4)

        loss = 0
        for pred in preds:
            trans_pred = pred * neg_inds + (1 - pred) * pos_inds
            weight = neg_weights * neg_inds + pos_inds
            all_loss = torch.log(1 - trans_pred) * torch.pow(trans_pred, 2) * weight
            all_loss = all_loss.sum()

            if num_pos > 0:
                all_loss /= num_pos

            loss -=  all_loss
        return loss

**Comment by xingyizhou** [comment]
Hi, 
  Thanks very much for the feedback of the focal loss implementation. It can be extremely helpful to this project. However, according to my test, it is not significantly faster than my original implementation (<1%). How faster is it in your environment?

**Comment by moothes** [comment]
Well, I test it with original CornerNet code. My code is:

    def _neg_loss(preds, gt):
        pos_inds = gt.ge(1.).float()
        neg_inds = gt.lt(1.).float()

        num_pos  = pos_inds.float().sum()
        neg_weights = torch.pow(1 - gt, 4)

        loss = 0
        for pred in preds:
            st = time.time()
        
            trans_pred = pred * neg_inds + (1 - pred) * pos_inds
            weight = neg_weights * neg_inds + pos_inds
            all_loss = torch.log(1 - trans_pred) * torch.pow(trans_pred, 2) * weight
            all_loss = all_loss.sum()

            if num_pos > 0:
                all_loss /= num_pos
            print('  my neg_loss : {}, using time: {}.'.format(all_loss, time.time() - st))

            st = time.time()
        
            pos_loss = torch.log(pred) * torch.pow(1 - pred, 2) * pos_inds
            neg_loss = torch.log(1 - pred) * torch.pow(pred, 2) * neg_weights * neg_inds

            pos_loss = pos_loss.sum()
            neg_loss = neg_loss.sum()

            all_loss = pos_loss + neg_loss
            if num_pos > 0:
                all_loss = all_loss / num_pos

            print('your neg_loss : {}, using time: {}.\n'.format(all_loss, time.time() - st))

            loss -=  all_loss
        return loss

And the results are:
![a](https://user-images.githubusercontent.com/17693547/54177139-279b0a00-44cc-11e9-9871-48e22d4197d4.JPG)

I think it's slightly faster than your implementation. Maybe not significantly, but it must be faster.

**Comment by moothes** [comment]
Thank you for your reminding. I add `torch.cuda.synchronize()` before print:

        torch.cuda.synchronize()
        print('  my neg_loss : {}, using time: {}.'.format(all_loss, time.time() - st))

        torch.cuda.synchronize()
        print('your neg_loss : {}, using time: {}.\n'.format(all_loss, time.time() - st))

And results are:
![a](https://user-images.githubusercontent.com/17693547/54180479-bd3b9700-44d6-11e9-97e8-8fc826c08fcc.JPG)
